Adding a emoji support

GAMELASTER opened the pull request:
As a suggestion of #232, I did it. The sizes available are a `small` and `medium`, big isn't working.
It's available a one more size (a extra small), but I don't found a name of variable. (you can "spawn" only two sizes of emoji).

The "like" :+1: button isn't a emoji, it's a sticker, where
`369239263222822` is a small
`369239343222814` is a medium
`369239383222810` is a large

bsansouci commented:
Thanks for contributing. In the future could you not make a PR with two different changes? That would help us.

The sound part isn't quite right. Looking at #221 @ICSandy's change just uploads it as an audio file. We looked into it and committed this 9b2a5d3bfd05e5141165a9a565673ce7a23282ef.

For the emoji part, there are a ton of problems that need to be addressed before we can move forward. 
- some unicode emojis don't work on web and show this warning
<img width="671" alt="screen shot 2016-04-03 at 20 25 38" src="https://cloud.githubusercontent.com/assets/4534692/14238807/f90cc33e-f9e9-11e5-952b-aab7793f9ada.png">
- as you said, some emojis' large/medium look the same
- thumbs up is a sticker, not an emoji, so what happens when you try to give it a size? (Nothing happens)
- what happens when you try to send an emoji and an attachment? Should you be able to? (Answer: it works, and you should probably not be able

Also this function signature is really not great, as it closely follows what FB is doing. Can you change it to be `api.sendMessage({emoji: "🖕", emojiSize: "small"})`. The function should check that the keys `emoji` and `emojiSize` are _always_ provided together (and only provided together). It should also check that `emojiSize` is either `"small"`, `"medium"` or `"large"` and return an error if not. Lastly this should check that `emoji` is exactly one unicode character long (to prevent people from sending stickers instead).

GAMELASTER commented:
Sorry for inactive, im going to change it soon.

GAMELASTER commented:
@bsansouci 
So, I reverted my edits for not a merge problem and pushed it again.
I removed audio_ids patch.
To an emoji:
- The size of Like Sticker is about an Sticker ID, a big Like got another ID than a medium etc. So the sizing of the emoji don't do anything.
- With medium and large differences I think it's not a big problem
- Sending an attachment with emoji works nice, Example:

![image](https://cloud.githubusercontent.com/assets/1665373/14405412/f8669cda-fe8e-11e5-8657-a276704c30b5.png)
- Im little confused about throwing an errors in Handlers, I found one usage of error in handleUrl, but I dont think its a best way, maybe throwing an exception
- I have list of (almost) supported Facebook emojis, but I don't know if its a best idea to check if its valid emoji

Schmavery commented on src/sendMessage.js:
```diff
@@ -182,6 +184,26 @@ module.exports = function(defaultFuncs, api, ctx) {
+      {
+        return callback({error: "emojiSize property is empty"});
+      }
+      if(msg.emojiSize != "mini" && msg.emojiSize != "medium" && msg.emojiSize != "large")
```
mini?..

GAMELASTER replied:
Sorry! I again pushed a wrong revision of code!

Schmavery commented on src/sendMessage.js:
```diff
@@ -182,6 +184,26 @@ module.exports = function(defaultFuncs, api, ctx) {
+    if (msg.emoji) {
+      if(msg.emojiSize == null)
+      {
+        return callback({error: "emojiSize property is empty"});
```
What about people who set emojiSize but not emoji?

Schmavery commented on src/sendMessage.js:
```diff
@@ -182,6 +184,26 @@ module.exports = function(defaultFuncs, api, ctx) {
+      {
+        return callback({error: "emojiSize property is invalid"});
+      }
+      if(msg.emoji.length > 3)
```
Can we do a little better than this?

Schmavery commented on src/sendMessage.js:
```diff
@@ -182,6 +184,26 @@ module.exports = function(defaultFuncs, api, ctx) {
+      {
+        return callback({error: "Invalid emoji"});
+      }
+      form['message_batch[0][body]'] = msg.emoji;
```
Won't this silently overwrite the body of a message?  I'd like to give an error or at least a warning if someone tries to do this.

Schmavery commented:
Hey @GAMELASTER,
I'm not sure how you're getting this nice example of emoji+attachment.  It seems like when @bsansouci and I were trying it the other day, we had a lot of rendering issues in messenger.com and facebook.com
I agree that I wouldn't want to explicitly name every emoji that is supported by facebook in our code.
Just so it's clear, this explicitly doesn't support sending the standard thumbs up, is that correct?
Just some thoughts.

GAMELASTER commented:
Sorry for inactive! I will lookup this soon!

GAMELASTER commented:
Let's finish and close this one! Sorry for one year latency 🤣 

I tried to solve all the requests to change, there is some mine notices:
- Is hard to detect if the `emoji` really contains emoji
- If `emojiSize` is not set, it will send a classical size of emoji
- Is possible to send emoji (with all sizes) along with video and image (probably it will work on every other attachment)
- Another-skin emojis are working good too
- Default thumbs up emoji still works like sticker (every size is another sticker), but you can send a Thumbs Up emoji, but it looks like normal emoji

bsansouci commented on src/sendMessage.js:
```diff
@@ -184,6 +186,26 @@ module.exports = function(defaultFuncs, api, ctx) {
+      if (msg.emojiSize != "small" && msg.emojiSize != "medium" && msg.emojiSize != "large") {
+        return callback({error: "emojiSize property is invalid"});
+      }
+      if ((form['body'] != null || form['body'] != "") == false) {
```
This if statement seems a bit weird. You want to error when the body _is_ empty right?
So maybe do:
if (form['body'] != null && form['body'] != "") 
Because emojis are only valid when the body is empty

GAMELASTER replied:
Ah yes, you are right! Thanks and sorry, I did it at late-night, so my thinking has been little weird :-D I'm going change that.

bsansouci commented:
Seems good! I added just one comment.

GAMELASTER commented:
It's done! (forgot to write! :rofl: )

bsansouci commented:
Looks good! Thanks a lot 👍
